contracts/toCKB-typescript/src/switch/bonding/mod.rs

```diff
@@ -1,7 +1,113 @@
 use crate::switch::ToCKBCellDataTuple;
-use crate::utils::types::Error;
+use crate::utils::config::*;
+use crate::utils::tools::{get_xchain_kind, XChainKind};
+use crate::utils::types::{Error, ToCKBCellDataView};
+use bech32::{self, FromBase32};
+use ckb_std::ckb_constants::Source;
+use ckb_std::ckb_types::{bytes::Bytes, prelude::*};
+use ckb_std::debug;
+use ckb_std::high_level::{load_cell_capacity, load_cell_data, load_witness_args};
 use core::result::Result;
+use molecule::prelude::Vec;
 
-pub fn verify(_toCKB_data_tuple: &ToCKBCellDataTuple) -> Result<(), Error> {
+fn verify_data(
+    input_toCKB_data: &ToCKBCellDataView,
+    out_toCKB_data: &ToCKBCellDataView,
+) -> Result<u128, Error> {
+    let coin_kind = get_xchain_kind()?;
+    let amount: u128 = match coin_kind {
+        XChainKind::Btc => {
+            let (hrp, data) = bech32::decode(
+                core::str::from_utf8(out_toCKB_data.x_lock_address.as_ref()).unwrap(),
+            )
+            .map_err(|_| Error::XChainAddressInvalid)?;
+            if hrp != "bc" {
+                return Err(Error::XChainAddressInvalid);
+            }
+            let raw_data = Vec::<u8>::from_base32(&data).unwrap();
+            if raw_data.len() != 22 {
+                return Err(Error::XChainAddressInvalid);
+            }
+            if &raw_data[..2] != &[0x00, 0x14] {
+                return Err(Error::XChainAddressInvalid);
+            }
+            let btc_lot_size = out_toCKB_data.get_btc_lot_size()?;
+            btc_lot_size.get_sudt_amount()
+        }
+        XChainKind::Eth => {
+            if out_toCKB_data.x_lock_address.as_ref().len() != 20 {
+                return Err(Error::XChainAddressInvalid);
+            }
+            let eth_lot_size = out_toCKB_data.get_eth_lot_size()?;
+            eth_lot_size.get_sudt_amount()
+        }
+    };
+    if input_toCKB_data.user_lockscript.as_ref() != out_toCKB_data.user_lockscript.as_ref()
+        || input_toCKB_data.get_raw_lot_size() != out_toCKB_data.get_raw_lot_size()
+    {
+        return Err(Error::InvariantDataMutated);
+    }
+    Ok(amount)
+}
+
+fn verify_witness() -> Result<u8, Error> {
+    let witness_args = load_witness_args(0, Source::GroupInput)?.input_type();
+    if witness_args.is_none() {
+        return Err(Error::InvalidWitness);
+    }
+    let witness_bytes: Bytes = witness_args.to_opt().unwrap().unpack();
```

Comment:
I am not sure  if `witness_args.to_opt()` is None ?

Reply:
It is checked in line 55

Reply:
i thought if ```witness_args.to_opt()``` is none, then ```witness_args.is_none()``` should be none too